lib/solvers/TraceLabelOverlapAvoidanceSolver/TraceLabelOverlapAvoidanceSolver.ts

```diff
@@ -0,0 +1,649 @@
+    }
+
+    const unfriendlyOverlaps = overlaps.filter((o) => {
+      const originalNetIds = this.mergedLabelNetIdMap.get(
+        o.label.globalConnNetId,
+      )
+      if (originalNetIds) {
+        return !originalNetIds.has(o.trace.globalConnNetId)
+      }
+      return o.trace.globalConnNetId !== o.label.globalConnNetId
+    })
+
+    if (unfriendlyOverlaps.length === 0) {
+      this.solved = true
+      return
+    }
+
+    const updatedTracesMap = new Map<string, SolvedTracePath>()
+    for (const trace of this.traces) {
+      updatedTracesMap.set(trace.mspPairId, trace)
+    }
+
+    const processedTraceIds = new Set<string>()
+
+    for (const overlap of unfriendlyOverlaps) {
+      if (processedTraceIds.has(overlap.trace.mspPairId)) {
+        continue
+      }
+
+      const currentTraceState = updatedTracesMap.get(overlap.trace.mspPairId)!
+      const labelId = overlap.label.globalConnNetId
+      const detourCount = this.detourCountByLabel.get(labelId) || 0
+
+      const newTrace = rerouteCollidingTrace(
+        currentTraceState,
+        overlap.label,
+        this.problem,
+        this.PADDING_BUFFER,
+        detourCount,
+      )
+
+      if (newTrace.tracePath !== currentTraceState.tracePath) {
+        this.detourCountByLabel.set(labelId, detourCount + 1)
+      }
+
+      updatedTracesMap.set(currentTraceState.mspPairId, newTrace)
+      processedTraceIds.add(currentTraceState.mspPairId)
+    }
+
+    this.updatedTraces = Array.from(updatedTracesMap.values())
+
+    const minimizedTraces = minimizeTurnsWithFilteredLabels(
+      this.updatedTraces,
+      this.problem,
+      this.netTempLabelPlacements, // Use temp labels which include merged ones
+      this.mergedLabelNetIdMap,
+      this.PADDING_BUFFER,
+    )
+    if (minimizedTraces) {
+      this.updatedTraces = minimizedTraces
+    }
+
+    const balancedTraces = balanceLShapes(
+      this.updatedTraces,
+      this.problem,
+      this.netLabelPlacements,
+    )
+    if (balancedTraces) {
+      this.updatedTraces = balancedTraces
+    }
+
+    this.solved = true
+  }
+
+  override visualize(): GraphicsObject {
+    const graphics = visualizeInputProblem(this.problem)
+
+    if (!graphics.lines) graphics.lines = []
+    if (!graphics.circles) graphics.circles = []
+    if (!graphics.texts) graphics.texts = []
+    if (!graphics.rects) graphics.rects = []
+
+    for (const trace of Object.values(this.updatedTraces)) {
+      graphics.lines!.push({
+        points: trace.tracePath,
+        strokeColor: "purple",
+      })
+    }
+
+    for (const p of this.netLabelPlacements) {
+      graphics.rects!.push({
+        center: p.center,
+        width: p.width,
+        height: p.height,
+        fill: getColorFromString(p.globalConnNetId, 0.35),
+      })
+      graphics.points!.push({
+        x: p.anchorPoint.x,
+        y: p.anchorPoint.y,
+        color: getColorFromString(p.globalConnNetId, 0.9),
+      })
+    }
+
+    return graphics
+  }
+}
```

Comment:
try to break up this massive file by isolating the functions into the directory, you've done that for a couple files but i would do it for more, it's important for making things digestible for AI